game-core/src/main/java/games/strategy/triplea/delegate/battle/TargetGroup.java

```diff
@@ -0,0 +1,90 @@
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/** Group of firing units and their targets. */
+@Getter(AccessLevel.PRIVATE)
+public class TargetGroup {
+
+  private final Set<UnitType> firingUnitTypes;
+  private final Set<UnitType> targetUnitTypes;
+
+  public TargetGroup(final UnitType firingUnitType, final Set<UnitType> targetUnitTypes) {
+    firingUnitTypes = Sets.newHashSet(firingUnitType);
+    this.targetUnitTypes = targetUnitTypes;
+  }
+
+  public Collection<Unit> getFiringUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(firingUnitTypes));
+  }
+
+  public Collection<Unit> getTargetUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(targetUnitTypes));
+  }
+
+  /**
+   * Find all target groupings for firing units and enemy units based on canNotTarget and
+   * canNotBeTargetedBy attributes. Also takes into account if any destroyers are present which
+   * cancel canNotBeTargetedBy. Sort all the target groups so those with the least number of targets
+   * appear first.
+   */
+  public static List<TargetGroup> newTargetGroups(
+      final Collection<Unit> units, final Collection<Unit> enemyUnits) {
+
+    final Set<UnitType> unitTypes =
+        units.stream().map(unit -> unit.getType()).collect(Collectors.toSet());
+    final Set<UnitType> enemyUnitTypes =
+        enemyUnits.stream().map(unit -> unit.getType()).collect(Collectors.toSet());
+    final List<TargetGroup> targetGroups = new ArrayList<TargetGroup>();
+    for (final UnitType unitType : unitTypes) {
+      final Set<UnitType> targets = findTargets(unitType, unitTypes, enemyUnitTypes);
+      addToTargetGroups(unitType, targets, targetGroups);
+    }
+    return sortTargetGroups(targetGroups);
+  }
+
+  private static Set<UnitType> findTargets(
+      final UnitType unitType, final Set<UnitType> unitTypes, final Set<UnitType> enemyUnitTypes) {
+    final Set<UnitType> targets = new HashSet<>(enemyUnitTypes);
+    targets.removeAll(UnitAttachment.get(unitType).getCanNotTarget());
+    return unitTypes.stream().anyMatch(Matches.unitTypeIsDestroyer())
+        ? targets
+        : targets.stream()
+            .filter(
+                target -> !UnitAttachment.get(target).getCanNotBeTargetedBy().contains(unitType))
+            .collect(Collectors.toSet());
+  }
+
+  private static void addToTargetGroups(
+      final UnitType unitType, final Set<UnitType> targets, final List<TargetGroup> targetGroups) {
+    if (targets.isEmpty()) {
+      return;
+    }
+    targetGroups.stream()
```

Comment:
This looks strange, could you check this @ron-murhammer ? It looks like `targetGroups` is always empty.  Was the iteration supposed to be on `targets`?

Reply:
Its correct. It will be empty on the first call to this method but then future calls will be looking at existing TargetGroup in the list to see whether to add to that group or add a new TargetGroup.

Reply:
Ah, I see it now. It's a bit counter-intuitive for a loop or a stream to have an effect when the iterated item is empty.

It could be worthwhile IMO then, for explicitness/clarity, to special case this situation. IE:

```
if(targetGroups.isEmpty()) {
  targetGroiups.add(new TargetGroup(unitType, targets);
  return;
}

```

Though, I think it's now more clear to me how we could avoid the mutable 'in/out' parameter, IMHO it would be even better if we returned a value (and also the name of this method becomes a bit easier to refine), eg:

```

 private static Optional<TargetGroup>  findTargetInTargetGroups( final UnitType unitType, final Set<UnitType> targets, final List<TargetGroup> targetGroups) {
     return targetGroups.stream()
              .filter(targetGroup -> targetGroup.getTargetUnitTypes().equals(targets))
              .findAny()
}
```

Then the `for` loop where this is used becomes a bit more clear and the overall control-flow IMO I think is more evident:

```
    for (final UnitType unitType : unitTypes) {
      final Set<UnitType> targets = findTargets(unitType, unitTypes, enemyUnitTypes);
     targetGroups.add(
           findTargetInTargetGroups(unitType, targets, targetGroups).orElseGet(() -> new TargetGroup(unitType, targets)));
    }
```

WDYT?

Reply:
I'm fairly certain that isn't equivalent as it would add the same TargetGroup to the list multiple times. Now you could probably change to an ordered set instead of a list and I think that would then work.

So I generally agree with trying to avoid mutable parameters but this alternative seems harder to understand as you really shouldn't be adding to targetGroups if you find identical targets and instead add to an existing TargetGroup.

So probably if you want mutable parameters then you could do something like:
```
Optional<TargetGroup> targetGroup = findTargetInTargetGroups(unitType, targets, targetGroups);
if (targetGroup.isPresent()) {
// add unit type to existing target group
} else {
// create and add new target group to list
}
```

Reply:
Perhaps being more verbose is okay. I do think we need a change here as this was not easy to grok. Let me please clarify what I think are the problems so that we can be on the same page:

1. (biggest): The pattern of streaming on an empty list to then append to that list is a bit novel and surprising. I had to double check that it even worked. This perhaps could be solved by special casing the empty list scenario and appending, otherwise do the find.

2. The break up of:
```
for loop
   do something
   do more stuff
done
```
Is not necessarily helpful. The 'do more stuff' piece does not take all of the logic or even a complete step out of the for loop. To understand the for loop, you really have to jump down two methods to see what side-effects occur. Having the 'do more stuff' function be functional could fix this as you would have different levels of abstraction, rather than part of the for loop simply moved elsewhere (indirection as opposed to abstraction). This could also be solved by simply inlining, but I think we want an abstraction, ie: a method call to calculate a result that is then used in the for loop.

3. The mutable in parameter, this I think is probably more a symptom of the second problem.

With that stated, I think we'll at least be on the same page. Also for consideration, I've been wondering if we want a multimap instead of a list. The multimap would take care for us of the adding a new element if none is there, otherwise we append to an existing list. That perhaps could simplify/remove some of the helper methods.

Reply:
Updated. Not sure where else to take it. I think we generally agree on the drawbacks but they aren't necessarily easy to solve without adding in other drawbacks as the logic here is fairly complex (we aren't just making a CRUD app). We are also getting to the point of essentially debating the design of a few methods which while important, we've already taken a few passes on and this PR is pretty large and been open for a while. I think at some point we have to push forward and consider follow on PRs if there are some ideas on how to further improve it.

Reply:
Generally agree, some quick comments:
- tactical design should be reasonable and not add debt unless there is a really strong trade-off to be made. Quick follow-ups for cleanup are not really a good thing.
- large PRs will tend to be open for a long time as they cannot be reviewed quickly. The size of a PR will also lead to more areas of feedback and it's less efficient to come back and re-review. Consider it to another extent, if someone wanted to game the review system, then to get less review, just open larger PRs. I think the rule of thumb is about 30 minutes per 200 lines of code (https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/). Hence, at 600 lines, you're requiring a dedicated 90 minutes. Given our time limitations, that means at least multiple days for a first pass only. 

I think perhaps the better way to go would be more iterations, smaller changes, and decoupling updates. If for example the TargetGroup were all on its own, it probably could have been more efficiently reviewed and would not have blocked anything else. It is water under the bridge at this point though, and I also appreciate how much effort it takes to break up larger changes as I certainly have, multiple times, spent literally days breaking up 600-1000 lines PRs